### PR TITLE
Fixed #1873 -- Handled multi-valued query parameters in admin changelist filters.

### PR DESCRIPTION
**Pull request by sarahboyce**
While filters are fresh in my head, thought I would have a look at [this ticket](https://code.djangoproject.com/ticket/1873). :rocket:

This enables multi-select filters but it does not add a way to do this from the UI.
The ticket originally suggested a way to do it in the UI but later on, most people just wanted the query parameters to be handled. So I _feel_ like I can get away with not having it in the UI.

In some cases, I decided multi-selects do not make much sense and I don't handle it. These cases are `BooleanFieldListFilter`, `EmptyFieldListFilter`, `DateFieldListFilter` (partly because being able to now do this month or this year etc doesn't add much value) and all of the `isnull` options.

`SimpleListFilter` also does not handle it, this is because the user defined `queryset` method depends on `value` and if this changed from a string to a list, this would probably be a breaking change. If anyone felt passionatlely that they need this, they could overwrite the `value` method to return the list and then handle it.

Not sure if the release note is enough or if I should mention something in the docs also? Happy for any suggestions.
Then yeah let me know what you think :+1:

### CODE REVIEW

**Review by felixxm — COMMENTED** [review]
@sarahboyce Thanks :+1:

It seems that changing `params.get()` to `get_value_from_parameters()` is missing somewhere as the following request crashes with `DatabaseError`:
> "GET /admin/auth/user/?groups__id__exact=1,2"

image

Also, most of the changes is related with switching `request.GET.items()` to `request.GET.lists()`. Can we do this separately? For example: _"Refs #1873 -- Used GET.lists() in admin filters."_ without handling list of values in filters :thinking:

**Comment by sarahboyce** [comment]
@felixxm `GET /admin/auth/user/?groups__id__exact=1,2` should be `GET /admin/auth/user/?groups__id__exact=1&groups__id__exact=2` (or maybe `GET /admin/auth/user/?groups__id__in=1,2` which I think is already supported). It doesn't allow comma seperated stuff unless it ends with `__in`. 

> Also, most of the changes is related with switching request.GET.items() to request.GET.lists(). Can we do this separately? For example: "Refs https://github.com/django/django/pull/1873 -- Used GET.lists() in admin filters." without handling list of values in filters thinking

Yes I think I can :+1:

**Comment by felixxm** [comment]
> @felixxm `GET /admin/auth/user/?groups__id__exact=1,2` should be `GET /admin/auth/user/?groups__id__exact=1&groups__id__exact=2` (or maybe `GET /admin/auth/user/?groups__id__in=1,2` which I think is already supported). It doesn't allow comma seperated stuff unless it ends with `__in`.

Yes, but without this patch it's silently ignored. We shouldn't crash with `DatabaseError` on invalid filters.

**Comment by sarahboyce** [comment]
Ah I see - good catch :+1:

**Comment by sarahboyce** [comment]
Ok I think both things are sorted now :crossed_fingers:

File: django/contrib/admin/views/main.py [hunk 1]
```diff
@@ -48,6 +48,7 @@
     IS_FACETS_VAR,
     IS_POPUP_VAR,
     TO_FIELD_VAR,
+    PAGE_VAR,
```
Comment (felixxm) [inline django/contrib/admin/views/main.py:52]: Adding `PAGE_VAR` to `IGNORED_PARAMS` looks like on unrelated bugfix :thinking: Can you move it to the separate PR with a regression test?

Comment (sarahboyce) [reply django/contrib/admin/views/main.py:52]: Ah no sorry it's because things get deleted from `self.params` but not from the new `self.filter_params` so `PAGE_VAR` and `ERROR_VAR` shouldn't be needed in the `IGNORED_PARAMS` - I will push up a change

Comment (sarahboyce) [reply django/contrib/admin/views/main.py]: Updated [1e05ab9](https://github.com/django/django/pull/16621/commits/1e05ab98edf07f060c09f5d5c81738f7cc597183) :+1:

File: django/contrib/admin/utils.py [hunk 2]
```diff
@@ -54,10 +54,17 @@ def lookup_spawns_duplicates(opts, lookup_path):
     return False
 
 
+def get_value_from_parameters(parameters, key):
```
Comment (felixxm) [inline django/contrib/admin/utils.py:57]: Maybe `get_last_value...` or `get_single_value...`

File: django/contrib/admin/utils.py [hunk 3]
```diff
@@ -54,10 +54,17 @@ def lookup_spawns_duplicates(opts, lookup_path):
...
+    value = parameters.get(key)
+    return value[-1] if isinstance(value, list) else value
+
+
 def prepare_lookup_value(key, value, separator=","):
     """
     Return a lookup value prepared to be used in queryset filtering.
     """
+    if isinstance(value, list):
+        value = value[-1]
```
Comment (felixxm) [inline django/contrib/admin/utils.py]: I think we should handle `__in` with list of values.

**Review by felixxm — COMMENTED** [review]
@sarahboyce Thanks for updates :+1:

File: django/contrib/admin/utils.py (hunk 3 above)
Comment (sarahboyce) [reply django/contrib/admin/utils.py]: This is the first commit, I think the `__in` case looks like `["1,2,3"]` and then it gets changed in the next bit

File: django/contrib/admin/utils.py [hunk 4]
```diff
@@ -67,6 +74,16 @@ def prepare_lookup_value(key, value, separator=","):
     return value
 
 
+def build_q_object_from_lookup_parameters(parameters):
+    q_object = models.Q()
+    for param, param_item_list in parameters.items():
+        param_q_obj = models.Q(**{param: param_item_list[0]})
+        for item in param_item_list[1:]:
+            param_q_obj |= models.Q(**{param: item})
+        q_object &= param_q_obj
```
[the remark this answers is not in the document]
Comment (sarahboyce) [reply django/contrib/admin/utils.py]: [fa240c9](https://github.com/django/django/pull/16621/commits/fa240c94e22430086bc6788e9e274cdc6bb33e87)

File: django/contrib/admin/utils.py [hunk 5]
```diff
@@ -74,6 +74,16 @@ def prepare_lookup_value(key, value, separator=","):
...
```
[the remark this answers is not in the document]
Comment (sarahboyce) [reply django/contrib/admin/utils.py]: Right I thought about it and I was going to have something like this:
```
def build_q_object_from_lookup_parameters(parameters):
    if not parameters:
        return models.Q()
    field_q_objects = {}
    for param, param_item_list in parameters.items():
        field = param.split("__")[0]
        q_object = reduce(or_, (models.Q((param, item)) for item in param_item_list))
        if field in field_q_objects:
            field_q_objects[field] = field_q_objects[field] | q_object
        else:
            field_q_objects[field] = q_object
    return reduce(and_, field_q_objects.values())
```
But it breaks a bunch of tests because we have things like date ranges with `{"date_registered__gte": self.today, "date_registered__lt": self.tomorrow}` which were previously an `&` and this makes them an `|`

So I think different lookups in the same filter should be not be OR-ed

File: django/contrib/admin/views/main.py [hunk 6]
```diff
@@ -123,10 +123,13 @@ def __init__(
             )
         self.to_field = to_field
         self.params = dict(request.GET.items())
+        self.filter_params = dict(request.GET.lists())
```
Comment (felixxm) [inline django/contrib/admin/views/main.py:127]: `self.params` is now only use for ordering. Do you think it's worth keeping it? :thinking:
Maybe we could use this instead:
```python
self.order_params = request.GET[ORDER_VAR] if ORDER_VAR in request.GET else None
```

File: django/contrib/admin/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply django/contrib/admin/utils.py]: Agreed.

File: django/contrib/admin/views/main.py [hunk 7]
```diff
@@ -170,9 +174,12 @@ def get_filters(self, request):
...
 
-        for key, value in lookup_params.items():
-            if not self.model_admin.lookup_allowed(key, value):
-                raise DisallowedModelAdminLookup("Filtering by %s not allowed" % key)
+        for key, value_list in lookup_params.items():
+            for value in value_list:
+                if not self.model_admin.lookup_allowed(key, value):
+                    raise DisallowedModelAdminLookup(
+                        "Filtering by %s not allowed" % key
+                    )
```
Comment (felixxm) [inline django/contrib/admin/views/main.py:182]:
```suggestion
                    raise DisallowedModelAdminLookup(f"Filtering by {key} not allowed")
```

File: docs/releases/5.0.txt [hunk 8]
```diff
@@ -54,6 +54,9 @@ Minor features
 * The new :meth:`.AdminSite.get_log_entries` method allows customizing the
   queryset for the site's listed log entries.
 
+* The admin changelist handles multi-valued query parameters for ``CharField``,
```
Comment (felixxm) [inline docs/releases/5.0.txt:57]:
```
* The admin changelist now handles ...
```

File: django/contrib/admin/views/main.py (hunk 6 above)
Comment (sarahboyce) [reply django/contrib/admin/views/main.py:127]: It's also used in the `date_hierarchy` templatetag and a couple of `search_form.html` inserts.
I can still add `self.order_params` and try to migrate the stuff to the `filter_params` 

But I guess removing it could cause issues if anyone is using this in some overriding of inserts and stuff :thinking: what do you think?

Comment (felixxm) [reply django/contrib/admin/views/main.py:127]: > But I guess removing it could cause issues if anyone is using this in some overriding of inserts and stuff thinking what do you think?

As far as I'm aware, it is not documented, so we can theoretically remove it :shrug: , but fine, let's leave it :+1:

**Review by felixxm — APPROVED** [review]
@sarahboyce Many thanks :+1: I think it's ready :rocket: 
 
I moved test for `IncorrectLookupParameters` to a separate commit, added test for `AllValuesFieldListFilter`, and pushed edits to release notes.

**Comment by felixxm** [comment]
buildbot, test on selenium.
